### ISSUE TITLE
Env Plugin

### ISSUE THREAD

**Issue by shadowtime2000** [alephjs/aleph.js#78]
A plugin to replace `.env` variables in a certain manner. Initially proposed in #77. Initial design used regex to replace anything surrounded by handlebars. I think we should use `deno_swc` which will allow us to instead replace `Deno.env.` anything. We could possibly look into requiring `ALEPH_PUBLIC_` prefix to `.env` variables unless it is accessed inside a `useDeno` hook.

**Comment by shadowtime2000** [comment]
@ije Thoughts? @alfredosalzillo Are you still interested in creating such an env plugin?

**Comment by ije** [comment]
maybe too much to import the `deno_swc` for the `env` plugin, **0.3** will include a new compiler using swc i can implement this function without plugin imported.

**Comment by shadowtime2000** [comment]
@ije Makes sense!

**Comment by ije** [comment]
i perfer a new concept **swc-marco** can be a solution:

```js
import env from 'https://deno.land/x/aleph/swc-macros/env.macro.ts'

const value = env('KEY')
```

it will be transpiled to below code:

```js
const value = __ALEPH_ENV.get('KEY')
```

and the **macro** will export the `deno.env` as `__ALEPH_ENV`.

what do you think? @shadowtime2000 @alfredosalzillo

**Comment by ije** [comment]
with that, people can write their own macro that gives more power to define your code syntax.

**Comment by alfredosalzillo** [comment]
I agree. I prefer this solution too.

**Comment by ije** [comment]
for example, we can transform graphql string to AST at the build time with macro:

```js
import gql from 'https://deno.land/x/aleph/macros/graphql.macro.ts'

const query = gql`
  {
    user(id: 5) {
      firstName
      lastName
    }
  }
`
```

real code:

```js
const query = {
  "kind": "Document",
  "definitions": [...]
}
```

**Comment by shadowtime2000** [comment]
I like the idea of an SWC macro, it could possibly then help us with #60.

**Comment by shadowtime2000** [comment]
@ije How exactly would this be written? Will we write it within the new compiler after it is merged or will we another repo for this project or something? I think SWC macros could be useful just all around in a ton of places.

**Comment by ije** [comment]
it seems like:

```ts
// https://deno.land/x/aleph/swc-macros/env.macro.ts

export default function EnvMacro(key: string): string {
  return ''
}

EnvMacro.macro = {
  inject(): string {
    return `window. __ALEPH_ENV=${JSON.stringify(Deno.env)}`
  },
  transform(arg0): string {
    return `__ALEPH_ENV.get(${arg0})`
  }
}
```

**Comment by shadowtime2000** [comment]
@ije Thx, but I was actually asking about how the compiler which expands the macros will be organized. I don't think we should have an `inject` method which injects that into the main file, I think it would be better if we just did something like this:

```ts
export default function EnvMacro(key: string): string {
  return ''
}

EnvMacro.macro = {
	transform(arg0): string {
		return Deno.env.get(arg0);
	}
}
```

**Comment by ije** [comment]
@shadowtime2000 the `transform` should return a js code fragment that why i added the `inject` method(it's optional) to inject deno runtime variables. i will add the feature in the new compiler, no extra dependency needed.

**Comment by shadowtime2000** [comment]
Tbh, I feel like macros purely for a framework like this is a little overkill and we could look into creating a library for these macros that doesn't require Aleph to be used.

**Comment by shadowtime2000** [comment]
@ije I don't think it is a good idea to dump everything in `Deno.env` into a global namespace because there could be unused `.env` variables for stuff like API keys which could be used in the `/api/` routes, which makes that a security concern. I think this should be good enough:

```ts
export default function EnvMacro(key: string): string {
  return ''
}

EnvMacro.macro = {
	transform(arg0): string {
		return `"${Deno.env.get(arg0)}"`;
	}
}
```
So `env("foo")` will be replaced with `"bar"`.

**Comment by ije** [comment]
@shadowtime2000 you are right it's a bad idea to inject `deno.env`!

**Comment by ije** [comment]
i feel that the maco function is very easy to implement in the new compile with swc, if there is a statement like `import macroFn from "./xxx.macro.ts"`, the **macroFn** will be treated as a macro. if we decide to implement it in a library then we need an ast compiler that is heavy.

**Comment by shadowtime2000** [comment]
@ije Good point, it would be easier to make it a library once there is a more standard form for ast modification and compilation and stuff.

**Comment by ije** [comment]
@shadowtime2000 for sure

**Comment by shadowtime2000** [comment]
We should probably allow returning an AST or something for more organized complex macros.

**Comment by shadowtime2000** [comment]
I think we should maybe instead just invoke a macro if macro data is present on the function.
